pkg/actor/resize_pvc.go

```diff
@@ -0,0 +1,248 @@
+		log.Info(" resize pvc statefulset does not have all replicas up")
+		return NotReadyErr{Err: errors.New("resize pvc statefulset does not have all replicas up")}
+	}
+
+	// Maybe this should be an error since we should not have this, but I wanted to check anyways
+	if len(statefulSet.Spec.VolumeClaimTemplates) == 0 {
+		log.Info("Skipping PVC resize as PVCs do not exist")
+		return nil
+	}
+
+	clientset, err := kubernetes.NewForConfig(rp.config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create kubernetes clientset")
+	}
+
+	stsStorageSizeDeployed := statefulSet.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()
+	stsStorageSizeSet := cluster.Spec().DataStore.VolumeClaim.PersistentVolumeClaimSpec.Resources.Requests.Storage()
+
+	// If the sizes match do not resize
+	if stsStorageSizeDeployed.Equal(stsStorageSizeSet.DeepCopy()) {
+		log.Info("Skipping PVC resize as sizes match")
+		return nil
+	}
+
+	log.Info("Starting PVC resize")
+
+	// Find all of the PVCs and resize them
+	if err := rp.findAndResizePVC(ctx, statefulSet, cluster, clientset); err != nil {
+		return errors.Wrapf(err, "updating PVCs for statefulset %s.%s", cluster.Namespace(), cluster.StatefulSetName())
+	}
+
+	log.Info("Starting updating sts")
+
+	// Update the STS with the correct volume size, in case more pods are created
+	// We will create a copy and update the copy, and then delete the original without
+	// deleting the Pods.  The new sts is then used to create a new statefulset.
+	if err := rp.updateSts(ctx, statefulSet, cluster); err != nil {
+		return errors.Wrapf(err, "updating statefulset %s.%s", cluster.Namespace(), cluster.StatefulSetName())
+	}
+
+	if !cluster.Spec().DataStore.SupportsAutoResize {
+		log.Info("Starting rolling sts")
+		// Roll the entire STS in order for the Pods to resize
+		if err := rp.rollSts(ctx, cluster, clientset); err != nil {
+			return errors.Wrapf(err, "error restarting statefulset %s.%s", cluster.Namespace(), cluster.StatefulSetName())
+		}
+	} else {
+		log.Info("Volumes support autoresizing so not restarting STS Pods")
+	}
+
+	log.Info("PVC resize completed")
+	CancelLoop(ctx)
+
+	return nil
+}
+
+// updateSts updates the size of an STS' VolumeClaimTemplate to match the new size in the CR.
+// In order to update the volume claim template we have to delete the STS without cascading and then
+// create the sts.
+func (rp *resizePVC) updateSts(ctx context.Context, sts *appsv1.StatefulSet, cluster *resource.Cluster) error {
+
+	// delete the original sts, but do not delete the Pods
+	orphan := metav1.DeletePropagationOrphan
+	if err := rp.client.Delete(ctx, sts, &client.DeleteOptions{PropagationPolicy: &orphan}); err != nil {
+		return err
+	}
+
+	// Use same StatefulSetBuilder that we run in Deploy to
+	// rebuild and save the StatefulSet with the new PVC size
+	r := resource.NewManagedKubeResource(ctx, rp.client, cluster, kube.AnnotatingPersister)
+	_, err := (resource.Reconciler{
```

Comment:
I am getting:

```
2020-12-01T00:24:15.372Z	ERROR	controller.CrdbCluster	action failed	{"CrdbCluster": "default/cockroachdb", "error": "updating statefulset default.cockroachdb: failed to reconcile statefulset: StatefulSet.apps \"cockroachdb\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden", "errorVerbose": "updating statefulset default.cockroachdb: failed to reconcile statefulset: StatefulSet.apps \"cockroachdb\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden\n(1) attached stack trace\n  -- stack trace:\n  | github.com/cockroachdb/cockroach-operator/pkg/actor.(*resizePVC).Act\n  | \tpkg/actor/resize_pvc.go:126\n  | [...repeated from below...]\nWraps: (2) updating statefulset default.cockroachdb\nWraps: (3) attached stack trace\n  -- stack trace:\n  | github.com/cockroachdb/cockroach-operator/pkg/actor.(*resizePVC).updateSts\n  | \tpkg/actor/resize_pvc.go:170\n  | github.com/cockroachdb/cockroach-operator/pkg/actor.(*resizePVC).Act\n  | \tpkg/actor/resize_pvc.go:125\n  | github.com/cockroachdb/cockroach-operator/pkg/controller.(*ClusterReconciler).Reconcile\n  | \tpkg/controller/cluster_controller.go:101\n  | sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n  | \texternal/io_k8s_sigs_controller_runtime/pkg/internal/controller/controller.go:256\n  | sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n  | \texternal/io_k8s_sigs_controller_runtime/pkg/internal/controller/controller.go:232\n  | sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n  | \texternal/io_k8s_sigs_controller_runtime/pkg/internal/controller/controller.go:211\n  | k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n  | \texternal/io_k8s_apimachinery/pkg/util/wait/wait.go:155\n  | k8s.io/apimachinery/pkg/util/wait.BackoffUntil\n  | \texternal/io_k8s_apimachinery/pkg/util/wait/wait.go:156\n  | k8s.io/apimachinery/pkg/util/wait.JitterUntil\n  | \texternal/io_k8s_apimachinery/pkg/util/wait/wait.go:133\n  | k8s.io/apimachinery/pkg/util/wait.Until\n  | \texternal/io_k8s_apimachinery/pkg/util/wait/wait.go:90\n  | runtime.goexit\n  | \tsrc/runtime/asm_amd64.s:1373\nWraps: (4) failed to reconcile statefulset\nWraps: (5) StatefulSet.apps \"cockroachdb\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden\nError types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *errors.StatusError"}
```

Here.

Reply:
I think we talked about this? but you have to poll to ensure the sts is fully deleted before recreating it.

The foreground propagation policy is mutually exclusive with orphan for some reason.

Reply:
Yep, we talked about it.  I have a retry with a backoff implemented now.